### PR TITLE
sync clusterdeployment label by cluster

### PR DESCRIPTION
**Pull request by ldpliu**
https://github.com/open-cluster-management/backlog/issues/14072

### CODE REVIEW

**Comment by ldpliu** [comment]
please do not merge this, as we need to change the validatingwebhook config first in installer.

File: deploy/foundation/hub/resources/webhook/webhook-validating-config.yaml [hunk 1]
```diff
@@ -29,7 +29,6 @@ webhooks:
         apiVersions:
           - "v1"
         resources:
-          - clusterdeployments
```
Comment (qiujian16) [inline deploy/foundation/hub/resources/webhook/webhook-validating-config.yaml:32]: why this is removed?

File: pkg/controllers/clusterset/managedclusterlabel/managedclusterlabel_controller.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// This controller sync the managedcluster's clusterSet Label with releated clusterdeployment's ClusterSet only when the clusterdeployment is claimed from pool.
```
Comment (qiujian16) [inline pkg/controllers/clusterset/managedclusterlabel/managedclusterlabel_controller.go:28]: shouldn't it has been handled by clusterdeployment_controller?

File: pkg/webhook/clusterset/validatingWebhook.go [hunk 3]
```diff
@@ -176,34 +153,79 @@ func (a *AdmissionHandler) validateCreateRequest(request *v1.AdmissionRequest) *
...
-// 2. the user must have managedClusterSet/join <clusterSet-A/clusterSet-B> permission if update resource to change
-// the managedClusterSet label from clusterSet-A to clusterSet-B.
-// 3. the user must have managedClusterSet/join <all> permission if update resource to add/remove the managedClusterSet label.
+// 2. For clusterpool, do not allow to update managedClusterSet label
+// 3. For managedClusters, when their is no clusterdeployment (import a existing cluster), or the clusterdeployment is not
+//    refer to a clusterpool(directlly create a cluster), we need to check if he has permission to update the managedClusterSet label.
+//   3.1. the user must have managedClusterSet/join <clusterSet-A/clusterSet-B> permission if update resource to change
+//         the managedClusterSet label from clusterSet-A to clusterSet-B.
+//   3.2. the user must have managedClusterSet/join <all> permission if update resource to add/remove the managedClusterSet label.
+// 4. if the managedcluster claimed from a clusterpool, we will always sync the managedcluster clusterset by clusterdeployment.
```
Comment (qiujian16) [inline pkg/webhook/clusterset/validatingWebhook.go:162]: this is confusing...would rephrase the comment here?

File: deploy/foundation/hub/resources/webhook/webhook-validating-config.yaml (hunk 1 above)
Comment (ldpliu) [reply deploy/foundation/hub/resources/webhook/webhook-validating-config.yaml:32]: Currentlly, we will always sync clusterdeployments clusterset label by clusterpool or managedcluster's clusterset. 
So do not need to use webhook to check it anymore.

File: pkg/controllers/clusterset/managedclusterlabel/managedclusterlabel_controller.go (hunk 2 above)
Comment (ldpliu) [reply pkg/controllers/clusterset/managedclusterlabel/managedclusterlabel_controller.go:28]: clusterdeployment _controller used to add/remove clusterset label for clusterdeployment. 
this controller used to sync the clusterdeployment clusterset to managedcluster.

File: pkg/webhook/clusterset/validatingWebhook.go (hunk 3 above)
Comment (ldpliu) [reply pkg/webhook/clusterset/validatingWebhook.go]: done

File: pkg/controllers/clusterset/clusterdeployment/clusterdeployment_controller.go [hunk 4]
```diff
@@ -23,8 +24,9 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// This controller sync the clusterdeployment's utils.ClusterSetLabel with releated clusterpool's utils.ClusterSetLabel
-// if the clusterpool did not exist, do nothing.
+// This controller sync the clusterdeployment's ClusterSet Label
+// 1. if it's created directlly(create a cluster), we will sync it's clusterset label by managedcluster
+// 2. if it's created by clusterpool, we will sync the clusterset label by clusterpool
```
Comment (qiujian16) [inline pkg/controllers/clusterset/clusterdeployment/clusterdeployment_controller.go:29]: this needs more explanation about the reasoning.

File: pkg/webhook/clusterset/validatingWebhook.go [hunk 5]
```diff
@@ -175,35 +172,91 @@ func (a *AdmissionHandler) validateCreateRequest(request *v1.AdmissionRequest) *
...
+
+		// Get managedcluster's clusterdeployment
+		// 1. No related clusterdeployment, check if it has permission to update this cluster.
+		// 2. The related clusterdeployment is created from clusterpool, do not allow to update this cluster.
+		// 3. The related clusterdeployment is not created from clusterpool, check if it has permission to update this cluster.
+		clusterDeployment, err := a.HiveClient.HiveV1().ClusterDeployments(managedCluster.Name).Get(context.TODO(), managedCluster.Name, metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			rejectStatus.Result.Message = err.Error()
+			return rejectStatus
+		}
```
Comment (qiujian16) [inline pkg/webhook/clusterset/validatingWebhook.go:225]: you need to check if err != nil, right?

File: pkg/webhook/clusterset/validatingWebhook.go [hunk 6]
```diff
@@ -175,35 +172,91 @@ func (a *AdmissionHandler) validateCreateRequest(request *v1.AdmissionRequest) *
...
+			rejectStatus.Result.Message = "Do not allow to update claimed cluster's clusterset."
+			return rejectStatus
+		}
+		break
+	case clusterPoolsGVR:
+		// do not allow to update clusterpool's clusterset
+		rejectStatus.Result.Message = "Do not allow to update clusterpool's clusterset label"
+		return rejectStatus
+	case clusterDeploymentsGVR:
+		// check if the user has permission to update the clusterset.
```
Comment (qiujian16) [inline pkg/webhook/clusterset/validatingWebhook.go:238]: why cluster deployment label is allowed to update if there is pool ref.

File: pkg/controllers/clusterset/clusterdeployment/clusterdeployment_controller.go [hunk 7]
```diff
@@ -113,25 +154,41 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+			&managedcluster,
+		)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return ctrl.Result{}, nil
+			}
+			return ctrl.Result{}, err
 		}
-		return ctrl.Result{}, err
+		targetLabels = managedcluster.Labels
```
Comment (qiujian16) [inline pkg/controllers/clusterset/clusterdeployment/clusterdeployment_controller.go:174]: sync all labels or just set label?

File: pkg/controllers/clusterset/clusterdeployment/clusterdeployment_controller.go [hunk 8]
```diff
@@ -113,25 +154,41 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+		// clusterpool's clusterdeployment, update clusterdeployment label by clusterpool
+		clusterpool := &hivev1.ClusterPool{}
+		err = r.client.Get(ctx, types.NamespacedName{Namespace: clusterdeployment.Spec.ClusterPoolRef.Namespace, Name: clusterdeployment.Spec.ClusterPoolRef.PoolName}, clusterpool)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return ctrl.Result{}, nil
+			}
+			return ctrl.Result{}, err
+		}
+		targetLabels = clusterpool.Labels
```
Comment (qiujian16) [inline pkg/controllers/clusterset/clusterdeployment/clusterdeployment_controller.go:186]: sync all labels or just set label?

File: pkg/controllers/clusterset/clusterdeployment/clusterdeployment_controller.go (hunk 7 above)
Comment (ldpliu) [reply pkg/controllers/clusterset/clusterdeployment/clusterdeployment_controller.go:178]: it's just set label. 
In the blow func，we only sync the set filed
`utils.SyncMapFiled(&isModified, &clusterdeployment.Labels, targetLabels, **clustersetutils.ClusterSetLabel**)`

File: pkg/webhook/clusterset/validatingWebhook.go (hunk 6 above)
Comment (ldpliu) [reply pkg/webhook/clusterset/validatingWebhook.go:241]: In clusterdeployment controller, we need to add the set label to clusterdeployment follow pool's set.

**Review by qiujian16 — COMMENTED** [review]
/approve
